File: app/src/main/java/com/adsamcik/signalcollector/activities/FileSharingActivity.java
package com.adsamcik.signalcollector.activities;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.v4.content.FileProvider;
import android.util.SparseBooleanArray;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.adsamcik.signalcollector.R;
import com.adsamcik.signalcollector.utility.BottomSheetMenu;
import com.adsamcik.signalcollector.utility.Compress;
import com.adsamcik.signalcollector.utility.DataStore;

import java.io.File;
import java.util.ArrayList;

public class FileSharingActivity extends DetailActivity {
	private static final int SHARE_RESULT = 1;
	private static final String SHAREABLE_DIR_NAME = "shareable";
	File[] files;
	String[] fileNames;
	File shareableDir;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		files = getFilesDir().listFiles((dir1, name) -> name.startsWith(DataStore.DATA_FILE));
		if (files.length == 0) {
			TextView tv = new TextView(this);
			tv.setText(R.string.share_nothing_to_share);
			tv.setGravity(Gravity.CENTER_HORIZONTAL);
			createContentParent(true).addView(tv);
		} else {
			fileNames = new String[files.length];
			for (int i = 0; i < fileNames.length; i++)
				fileNames[i] = files[i].getName();

			LinearLayout parent = createContentParent(false);
			CoordinatorLayout layout = (CoordinatorLayout) ((ViewGroup) getLayoutInflater().inflate(R.layout.layout_file_share, parent)).getChildAt(parent.getChildCount() - 1);
			ListView listView = layout.findViewById(R.id.share_list_view);
			ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_multiple_choice, fileNames);
			listView.setAdapter(adapter);
			listView.setChoiceMode(AbsListView.CHOICE_MODE_MULTIPLE);

			View.OnClickListener shareOnClickListener = v -> {
				SparseBooleanArray sba = listView.getCheckedItemPositions();
				ArrayList<String> temp = new ArrayList<>();
				for (int i = 0; i < fileNames.length; i++)
					if (sba.get(i))
						temp.add(fileNames[i]);

				if (temp.size() == 0)
					Toast.makeText(this, R.string.share_nothing_to_share, Toast.LENGTH_SHORT).show();
				else {
					String[] arr = new String[temp.size()];
					temp.toArray(arr);
					File c = Compress.zip(files[0].getParent(), arr, "export_" + System.currentTimeMillis());
					assert c != null;
					File target = new File(c.getParent() + File.separatorChar + SHAREABLE_DIR_NAME + File.separatorChar + c.getName() + ".zip");
					shareableDir = new File(c.getParent() + File.separatorChar + SHAREABLE_DIR_NAME);
					if (shareableDir.exists() || shareableDir.mkdir()) {
						if (c.renameTo(target)) {
							Uri fileUri = FileProvider.getUriForFile(
									FileSharingActivity.this,
									"com.asdamcik.signalcollector.fileprovider",
									target);
							Intent shareIntent = new Intent();
							shareIntent.setAction(Intent.ACTION_SEND);
							shareIntent.putExtra(Intent.EXTRA_STREAM, fileUri);
							shareIntent.setType("application/zip");
							startActivityForResult(Intent.createChooser(shareIntent, getResources().getText(R.string.export_share_button)), SHARE_RESULT);

						}
					}

					target.deleteOnExit();
					shareableDir.deleteOnExit();
				}
			};

			BottomSheetMenu bottomSheetMenu = new BottomSheetMenu(layout);
			bottomSheetMenu.addItem(R.string.export_share_button, shareOnClickListener);

			bottomSheetMenu.addItem(R.string.select_all, (v) -> {
				for (int i = 0; i < fileNames.length; i++)
					listView.setItemChecked(i, true);
			});

			bottomSheetMenu.addItem(R.string.deselect_all, (v) -> {
				for (int i = 0; i < fileNames.length; i++)
					listView.setItemChecked(i, false);
			});

			bottomSheetMenu.showHide(750);
		}

		setTitle(R.string.export_share_button);
	}

	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);

		if (requestCode == SHARE_RESULT)
			DataStore.recursiveDelete(shareableDir);

		finish();
	}
}

File: app/src/main/java/com/adsamcik/signalcollector/activities/FileSharingActivity.kt
package com.adsamcik.signalcollector.activities

import android.content.Intent
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.content.FileProvider
import com.adsamcik.signalcollector.R
import com.adsamcik.signalcollector.components.BottomSheetMenu
import com.adsamcik.signalcollector.exports.IExport
import com.adsamcik.signalcollector.uitools.ColorView
import java.io.File
import java.util.zip.ZipFile

/**
 * Activity that allows user to share his collected data to other apps that support zip files
 */
class FileSharingActivity : DetailActivity() {
    private var shareableDir: File? = null
    private lateinit var root: ViewGroup

    private val files = ArrayList<IReadableFile>()


    private fun addAllFiles() {
        val storedFiles = LongTermStore.listFiles(this)

        storedFiles.forEach {
            val zipFile = ZipFile(it)
            val entryEnumeration = zipFile.entries()

            while (entryEnumeration.hasMoreElements()) {
                files.add(ReadableArchivedFile(zipFile, entryEnumeration.nextElement()))
            }
        }

        val temp = DataStore.getDir(this).listFiles { _, name -> name.startsWith(DataStore.DATA_FILE) || name.startsWith(DataStore.TMP_NAME) }
        files.addAll(temp.map { ReadableFile(it) })

        files.sortBy { it.time }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        addAllFiles()

        if (files.isEmpty()) {
            val tv = TextView(this)
            tv.setText(R.string.share_nothing_to_share)
            tv.gravity = Gravity.CENTER_HORIZONTAL
            root = createLinearContentParent(true)
            root.addView(tv)
        } else {
            val itemNames = files.map { it.name }
            root = createLinearContentParent(false)
            val layout = (layoutInflater.inflate(R.layout.layout_file_share, root) as ViewGroup).getChildAt(root.childCount - 1) as androidx.coordinatorlayout.widget.CoordinatorLayout
            val listView = layout.findViewById<ListView>(R.id.share_list_view)
            val adapter = ArrayAdapter(this, android.R.layout.simple_list_item_multiple_choice, itemNames)
            listView.adapter = adapter
            listView.choiceMode = AbsListView.CHOICE_MODE_MULTIPLE

            val shareOnClickListener = View.OnClickListener {
                val sba = listView.checkedItemPositions

                val selectedItems = files.filterIndexed { i, _ -> sba[i] }


                if (selectedItems.isEmpty())
                    Toast.makeText(this, R.string.share_nothing_to_share, Toast.LENGTH_SHORT).show()
                else {
                    val shareableDir = File(DataStore.getDir(this), SHAREABLE_DIR_NAME)
                    if (shareableDir.exists() || shareableDir.mkdir()) {
                        this.shareableDir = shareableDir
                        val exporterType = intent.extras!![EXPORTER_KEY] as Class<*>
                        val exporter = exporterType.newInstance() as IExport
                        val result = exporter.export(selectedItems, shareableDir)

                        if (result.isSuccessful) {
                            val fileUri = FileProvider.getUriForFile(
                                    this@FileSharingActivity,
                                    "com.adsamcik.signalcollector.fileprovider",
                                    result.file!!)
                            val shareIntent = Intent()
                            shareIntent.action = Intent.ACTION_SEND
                            shareIntent.putExtra(Intent.EXTRA_STREAM, fileUri)
                            shareIntent.type = result.mime
                            startActivityForResult(Intent.createChooser(shareIntent, resources.getText(R.string.export_share_button)), SHARE_RESULT)

                            result.file.deleteOnExit()
                        } else {
                            Toast.makeText(this, R.string.error_general, Toast.LENGTH_SHORT).show()
                        }

                        shareableDir.deleteOnExit()
                    }
                }
            }

            val bottomSheetMenu = BottomSheetMenu(layout)
            bottomSheetMenu.addItem(R.string.export_share_button, shareOnClickListener)

            bottomSheetMenu.addItem(R.string.select_all, View.OnClickListener {
                for (i in itemNames.indices)
                    listView.setItemChecked(i, true)
            })

            bottomSheetMenu.addItem(R.string.deselect_all, View.OnClickListener {
                for (i in itemNames.indices)
                    listView.setItemChecked(i, false)
            })


            root.post {
                bottomSheetMenu.showHide(1000)
            }

            colorManager!!.watchView(ColorView(bottomSheetMenu.menuRoot, 0))
        }

        setTitle(R.string.export_share_button)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == SHARE_RESULT)
            DataStore.recursiveDelete(shareableDir!!)

        finish()
    }

    companion object {
        private const val SHARE_RESULT = 1
        private const val SHAREABLE_DIR_NAME = "shareable"
        const val EXPORTER_KEY = "exporter"
    }
}
